# Warehouse-and-Retail-Sales
### URL DATASET →  https://catalog.data.gov/dataset/warehouse-and-retail-sales

### INFORMACIÓN ESTRUCTURA DE LOS DATOS → https://data.montgomerycountymd.gov/Community-Recreation/Warehouse-and-Retail-Sales/v76h-r7br/about_data

<img src="IMG\Columns.png">

In [314]:
import pandas as pd
import numpy as np
import random as rd
df = pd.read_csv('Warehouse_and_Retail_Sales.csv')
df

YEAR  MONTH                           SUPPLIER ITEM CODE  \
0       2020      1  REPUBLIC NATIONAL DISTRIBUTING CO    100009   
1       2020      1                          PWSWN INC    100024   
2       2020      1            RELIABLE CHURCHILL LLLP      1001   
3       2020      1          LANTERNA DISTRIBUTORS INC    100145   
4       2020      1               DIONYSOS IMPORTS INC    100293   
...      ...    ...                                ...       ...   
307640  2020      9                           DOPS INC     97896   
307641  2020      9                 ANHEUSER BUSCH INC     97918   
307642  2020      9                       HEINEKEN USA     97942   
307643  2020      9            RELIABLE CHURCHILL LLLP     97950   
307644  2020      9            RELIABLE CHURCHILL LLLP     97969   

                                   ITEM DESCRIPTION ITEM TYPE  RETAIL SALES  \
0                               BOOTLEG RED - 750ML      WINE          0.00   
1                         MOMENT DE PLAISIR - 750ML      WINE          0.00   
2               S SMITH ORGANIC PEAR CIDER - 18.7OZ      BEER          0.00   
3                     SCHLINK HAUS KABINETT - 750ML      WINE          0.00   
4                    SANTORINI GAVALA WHITE - 750ML      WINE          0.82   
...                                             ...       ...           ...   
307640  ST PETERS ORGANIC ENG ALE NR 12/CS - 16.9OZ      BEER          0.00   
307641               STELLA ARTOIS 2/12 NR - 11.2OZ      BEER        372.45   
307642                       TECATE 4/6 LNNR - 12OZ      BEER          7.79   
307643     S SMITH WINTER WELCOME NR 12/CS - 18.7OZ      BEER          0.00   
307644          S SMITH WINTER WELCOME 4/6NR - 12OZ      BEER          0.00   

        RETAIL TRANSFERS  WAREHOUSE SALES  
0                    0.0             2.00  
1                    1.0             4.00  
2                    0.0             1.00  
3                    0.0             1.00  
4                    0.0             0.00  
...                  ...              ...  
307640               0.0             1.00  
307641             315.0          3586.88  
307642               0.0             4.00  
307643               0.0             2.00  
307644               0.0             1.00  

[307645 rows x 9 columns]

#### ENTENDER CÓMO ES LA ESTRUCUTRA DE LA DATASET (TIPO DE VARIABLES, CANTIDAD DE DATOS Y COLUMNAS)

In [315]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307645 entries, 0 to 307644
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   YEAR              307645 non-null  int64  
 1   MONTH             307645 non-null  int64  
 2   SUPPLIER          307478 non-null  object 
 3   ITEM CODE         307645 non-null  object 
 4   ITEM DESCRIPTION  307645 non-null  object 
 5   ITEM TYPE         307644 non-null  object 
 6   RETAIL SALES      307642 non-null  float64
 7   RETAIL TRANSFERS  307645 non-null  float64
 8   WAREHOUSE SALES   307645 non-null  float64
dtypes: float64(3), int64(2), object(4)
memory usage: 21.1+ MB


#### INFORMACIÓN GENERAL DE LA DATASET (PROMEDIO; DESVIACIÖN ESTÁNDAR, MÍNIMO, MÁXIMO, ETC..)

In [316]:
df.describe()

YEAR          MONTH   RETAIL SALES  RETAIL TRANSFERS  \
count  307645.000000  307645.000000  307642.000000     307645.000000   
mean     2018.438525       6.423862       7.024071          6.936465   
std         1.083061       3.461812      30.986238         30.237195   
min      2017.000000       1.000000      -6.490000        -38.490000   
25%      2017.000000       3.000000       0.000000          0.000000   
50%      2019.000000       7.000000       0.320000          0.000000   
75%      2019.000000       9.000000       3.267500          3.000000   
max      2020.000000      12.000000    2739.000000       1990.830000   

       WAREHOUSE SALES  
count    307645.000000  
mean         25.294597  
std         249.916798  
min       -7800.000000  
25%           0.000000  
50%           1.000000  
75%           5.000000  
max       18317.000000

#### CONOCER CUANTOS ELEMENTOS NO TIENEN VALOR POR COLUMNA

In [317]:
df.isna().sum()

YEAR                  0
MONTH                 0
SUPPLIER            167
ITEM CODE             0
ITEM DESCRIPTION      0
ITEM TYPE             1
RETAIL SALES          3
RETAIL TRANSFERS      0
WAREHOUSE SALES       0
dtype: int64

ES INCORRECTO LOS VALORES PASADOS, YA QUE HAY MÁS VALORES EN CERO POR COLUMNA.

#### VAMOS A REVOLVER LA DATASET

In [318]:
# REVOLVER VALORES O SHUFFLED
df=df.iloc[np.random.permutation(df.index)].reset_index(drop=True)
df

YEAR  MONTH                         SUPPLIER ITEM CODE  \
0       2019      3           SUTTER HOME WINERY INC     80208   
1       2019      3           CASTLE BRANDS USA CORP     39640   
2       2017     12        SURVILLE ENTERPRISES CORP    341047   
3       2020      7                   THE WINE GROUP     76910   
4       2019      4                      LEGENDS LTD     63007   
...      ...    ...                              ...       ...   
307640  2017     12            BACCHUS IMPORTERS LTD    234820   
307641  2019      9            TI BEVERAGE GROUP LTD     10790   
307642  2017     12                      LEGENDS LTD     90972   
307643  2018      2  MICHAEL R DOWNEY SELECTIONS INC    195154   
307644  2017     12                JOS VICTORI WINES    311326   

                                 ITEM DESCRIPTION ITEM TYPE  RETAIL SALES  \
0                FOLIE A DEUX DRY CRK ZIN - 750ML      WINE          0.33   
1       GOSLING DK & STORMY W/GINGER BEER - 750ML    LIQUOR          0.68   
2                     EL GUARDIAN CRIANZA - 750ML      WINE          0.00   
3                           FISH EYE CHARD - 1.5L      WINE          7.27   
4            SOUTHERN TIER 2XVARIETY PACK 2/12 NR      BEER          0.00   
...                                           ...       ...           ...   
307640           KAPNOS PRIVATE LABEL RED - 750ML      WINE          0.00   
307641                        VAMPIRE CAB - 750ML      WINE          0.08   
307642                OKOCIM OK NR 20/CS - 16.9OZ      BEER          4.50   
307643        VIETTI BARBERA D'ASTI VIGNE - 750ML      WINE          0.16   
307644                 IN SITU RESA S/BLC - 750ML      WINE          0.64   

        RETAIL TRANSFERS  WAREHOUSE SALES  
0                   0.00              0.0  
1                   0.00              0.0  
2                   0.00              2.0  
3                   7.83             38.0  
4                   0.00              2.0  
...                  ...              ...  
307640              0.00              2.0  
307641              0.00              0.0  
307642              6.00              0.0  
307643              0.00              1.0  
307644              0.00             15.0  

[307645 rows x 9 columns]

#### VAMOS EMPEZAR A ELIMINAR COLUMNAS QUE NO APORTEN INFORMACIÓN RELEVANTE

In [319]:
df['ITEM CODE'].mode()

0       100293
1       100749
2        10103
3       101532
4        10197
         ...  
4591     99740
4592     99753
4593      9997
4594     99970
4595        WC
Name: ITEM CODE, Length: 4596, dtype: object

In [320]:
df=df.drop('ITEM CODE', axis=1)
df

YEAR  MONTH                         SUPPLIER  \
0       2019      3           SUTTER HOME WINERY INC   
1       2019      3           CASTLE BRANDS USA CORP   
2       2017     12        SURVILLE ENTERPRISES CORP   
3       2020      7                   THE WINE GROUP   
4       2019      4                      LEGENDS LTD   
...      ...    ...                              ...   
307640  2017     12            BACCHUS IMPORTERS LTD   
307641  2019      9            TI BEVERAGE GROUP LTD   
307642  2017     12                      LEGENDS LTD   
307643  2018      2  MICHAEL R DOWNEY SELECTIONS INC   
307644  2017     12                JOS VICTORI WINES   

                                 ITEM DESCRIPTION ITEM TYPE  RETAIL SALES  \
0                FOLIE A DEUX DRY CRK ZIN - 750ML      WINE          0.33   
1       GOSLING DK & STORMY W/GINGER BEER - 750ML    LIQUOR          0.68   
2                     EL GUARDIAN CRIANZA - 750ML      WINE          0.00   
3                           FISH EYE CHARD - 1.5L      WINE          7.27   
4            SOUTHERN TIER 2XVARIETY PACK 2/12 NR      BEER          0.00   
...                                           ...       ...           ...   
307640           KAPNOS PRIVATE LABEL RED - 750ML      WINE          0.00   
307641                        VAMPIRE CAB - 750ML      WINE          0.08   
307642                OKOCIM OK NR 20/CS - 16.9OZ      BEER          4.50   
307643        VIETTI BARBERA D'ASTI VIGNE - 750ML      WINE          0.16   
307644                 IN SITU RESA S/BLC - 750ML      WINE          0.64   

        RETAIL TRANSFERS  WAREHOUSE SALES  
0                   0.00              0.0  
1                   0.00              0.0  
2                   0.00              2.0  
3                   7.83             38.0  
4                   0.00              2.0  
...                  ...              ...  
307640              0.00              2.0  
307641              0.00              0.0  
307642              6.00              0.0  
307643              0.00              1.0  
307644              0.00             15.0  

[307645 rows x 8 columns]

SE VA ELIMINAR ITEM CODE DEBIDO A QUE SON CÓDIGOS QUE SE REPITEN VARIADAMENTE, NO HAY UNA TENDENCIA 

In [321]:
R=df['ITEM DESCRIPTION'].mode() # CONOCER LA MODA
G=df['ITEM DESCRIPTION'].nunique() # CONOCER CUANTOS ELEMENTOS ÚNICOS HAY EN LA COLUMNA
print(f''' 
    Moda {R}
    Total de elementos únicos {G} vs total {df['ITEM DESCRIPTION'].count()} 
''')

 
    Moda 0    BURGANS ALBARINO - 750ML
Name: ITEM DESCRIPTION, dtype: object
    Total de elementos únicos 34822 vs total 307645 



In [322]:
d =df[df['ITEM TYPE'].isna()].index
df = df.drop(d)
df.isna().sum()

YEAR                  0
MONTH                 0
SUPPLIER            167
ITEM DESCRIPTION      0
ITEM TYPE             0
RETAIL SALES          3
RETAIL TRANSFERS      0
WAREHOUSE SALES       0
dtype: int64

VAMOS EVALUAR SUPPLIER 

In [323]:
d =df[df['SUPPLIER'].isna()].index
for i in d.values:
    row = df.loc[i]
    # Contar el número de elementos vacíos en esa fila
    num_empty_elements = row.isna().sum()
    if num_empty_elements >1:
        print(f'Número de elementos vacíos en la fila {i}: {num_empty_elements}')  # Salida: 2
print(df['SUPPLIER'].isna().sum())

Número de elementos vacíos en la fila 46622: 2
Número de elementos vacíos en la fila 152436: 2
Número de elementos vacíos en la fila 276421: 2
167


SON 167 ELEMENTOS SIN NOMBRE DE 300,000 APROXIMADAMENTE, 3 ELEMENTOS QUE CUENTAN CON DOS ELEMENTOS VACIOS, A ESTOS LOS VAMOS A ELIMINAR

In [324]:
for i in d.values:
    row = df.loc[i]
    # Contar el número de elementos vacíos en esa fila
    num_empty_elements = row.isna().sum()
    if num_empty_elements >1:
        print(f'Número de elementos vacíos en la fila {i}: {num_empty_elements}')  # Salida: 2
        df = df.drop([i])

print(df.isna().sum())


Número de elementos vacíos en la fila 46622: 2
Número de elementos vacíos en la fila 152436: 2
Número de elementos vacíos en la fila 276421: 2
YEAR                  0
MONTH                 0
SUPPLIER            164
ITEM DESCRIPTION      0
ITEM TYPE             0
RETAIL SALES          0
RETAIL TRANSFERS      0
WAREHOUSE SALES       0
dtype: int64


EVALUAR SUPPLIER Y CAMBIAR LOS 164 ELEMENTOS PARA NO PERDER MUESTRAS, EN CASO DE NO ENCONTRAR CORRELACIÓN ELEMINARLAS YA QUE REPRESENTAN APROXIMADAMENTE UN  0.05466 % DE LAS MUESTRA TOTALES 

In [325]:
# Identificar los valores que no se repiten en la columna 'SUPPLIER'
unique_values = df['SUPPLIER'].value_counts()[df['SUPPLIER'].value_counts() == 1].index.tolist()

# Obtener los índices de las filas con valores faltantes en la columna 'SUPPLIER'
missing_indices = df[df['SUPPLIER'].isna()].index

# Reemplazar los valores faltantes con valores aleatorios de unique_values
for idx in missing_indices:
    df.at[idx, 'SUPPLIER'] = rd.choice(unique_values)

print("DataFrame después de reemplazar los valores faltantes:")
print(df)

DataFrame después de reemplazar los valores faltantes:
        YEAR  MONTH                         SUPPLIER  \
0       2019      3           SUTTER HOME WINERY INC   
1       2019      3           CASTLE BRANDS USA CORP   
2       2017     12        SURVILLE ENTERPRISES CORP   
3       2020      7                   THE WINE GROUP   
4       2019      4                      LEGENDS LTD   
...      ...    ...                              ...   
307640  2017     12            BACCHUS IMPORTERS LTD   
307641  2019      9            TI BEVERAGE GROUP LTD   
307642  2017     12                      LEGENDS LTD   
307643  2018      2  MICHAEL R DOWNEY SELECTIONS INC   
307644  2017     12                JOS VICTORI WINES   

                                 ITEM DESCRIPTION ITEM TYPE  RETAIL SALES  \
0                FOLIE A DEUX DRY CRK ZIN - 750ML      WINE          0.33   
1       GOSLING DK & STORMY W/GINGER BEER - 750ML    LIQUOR          0.68   
2                     EL GUARDIAN CRIANZA

SUSITUIMOS TODOS LOS ELEMENTOS FALTANTES POR ALGÚN ELEMENTO DE LA COLUMNA ALEATORIAMENTE

In [326]:
df.isna().sum()

YEAR                0
MONTH               0
SUPPLIER            0
ITEM DESCRIPTION    0
ITEM TYPE           0
RETAIL SALES        0
RETAIL TRANSFERS    0
WAREHOUSE SALES     0
dtype: int64

ELIMINAR EL ÚLTIMO DE LOS VALORES FALTANTE DE ITEM TYPE

In [327]:
d =df[df['ITEM TYPE'].isna()].index
df = df.drop(d)
df.isna().sum()

YEAR                0
MONTH               0
SUPPLIER            0
ITEM DESCRIPTION    0
ITEM TYPE           0
RETAIL SALES        0
RETAIL TRANSFERS    0
WAREHOUSE SALES     0
dtype: int64

CAMBIAREMOS LOS DATOS POR NUMEROS PARA PODER PROCESAR LA INFORMACIPON. UTILIZAREMOS LA CODIFICACION DE FRECUENCIA 

In [328]:
'''
Codificación de Frecuencia (Frequency Encoding): 
Codifica las categorías según la frecuencia con la que aparecen en el dataset. 
Esto es útil para reducir la dimensionalidad mientras conservas información sobre la distribución.
'''
freq_encoding = df['ITEM DESCRIPTION'].value_counts().to_dict()
df['ITEM DESCRIPTION'] = df['ITEM DESCRIPTION'].map(freq_encoding)

freq_encoding = df['SUPPLIER'].value_counts().to_dict()
df['SUPPLIER'] = df['SUPPLIER'].map(freq_encoding)


q=[]
# Obtiene los valor que no se repitan 
for i in df['ITEM TYPE'].values: 
    if i not in q:
        q.append(i)


item_dict = {item: index + 1 for index, item in enumerate( q)}
df['ITEM TYPE'] = df['ITEM TYPE'].map(item_dict)
df


YEAR  MONTH  SUPPLIER  ITEM DESCRIPTION  ITEM TYPE  RETAIL SALES  \
0       2019      3      2107                13          1          0.33   
1       2019      3       727                24          2          0.68   
2       2017     12       618                17          1          0.00   
3       2020      7      4037                24          1          7.27   
4       2019      4     13634                13          3          0.00   
...      ...    ...       ...               ...        ...           ...   
307640  2017     12      9150                 4          1          0.00   
307641  2019      9        93                19          1          0.08   
307642  2017     12     13634                24          3          4.50   
307643  2018      2       593                24          1          0.16   
307644  2017     12       905                24          1          0.64   

        RETAIL TRANSFERS  WAREHOUSE SALES  
0                   0.00              0.0  
1                   0.00              0.0  
2                   0.00              2.0  
3                   7.83             38.0  
4                   0.00              2.0  
...                  ...              ...  
307640              0.00              2.0  
307641              0.00              0.0  
307642              6.00              0.0  
307643              0.00              1.0  
307644              0.00             15.0  

[307641 rows x 8 columns]

HACEMOS SUBDIVISION DE ELEMENTOS DE TEST, TRAIN Y VALIDACIÓN 

In [329]:
# REVOLVER VALORES O SHUFFLED
# Establecer la semilla para reproducibilidad
np.random.seed(42)

# Mezclar el DataFrame
df= df.sample(frac=1, random_state=42).reset_index(drop=True)

In [330]:
# Calcular el tamaño de los conjuntos
total_size = len(df)
test_size = int(total_size * 0.15)
validation_size = int(total_size * 0.10)
train_size = total_size - test_size - validation_size

# Dividir en entrenamiento, validación y prueba
df_train = df[:train_size]
df_val = df[train_size:train_size + validation_size]
df_test = df[train_size + validation_size:]




In [331]:
# ENTRENAMIENTO 75%
# Separar características y variable objetivo
X_train = df_train.drop(columns=['RETAIL SALES'])
y_train = df_train['RETAIL SALES']


In [332]:
# VALIDACION 10% 
X_val = df_val.drop(columns=['RETAIL SALES'])
y_val = df_val['RETAIL SALES']

In [333]:
# TEST 15%
X_test = df_test.drop(columns=['RETAIL SALES'])
y_test = df_test['RETAIL SALES']


In [334]:
print(f"Tamaño del conjunto de entrenamiento: {X_train.shape[0]}")
print(f"Tamaño del conjunto de validación: {X_val.shape[0]}")
print(f"Tamaño del conjunto de prueba: {X_test.shape[0]}")

Tamaño del conjunto de entrenamiento: 230731
Tamaño del conjunto de validación: 30764
Tamaño del conjunto de prueba: 46146


CONTAR CUANTOS CEROS TENEMOS POR MUESTRA

In [335]:
for i in df.columns:
    x=np.sum(df[i] == 0.000)
    print(f'''{i} Valores con 0: {x} Valores totales: {df[i].count()} Ceros del total: {x*100/df[i].count()}%
            ''')

YEAR Valores con 0: 0 Valores totales: 307641 Ceros del total: 0.0%
            
MONTH Valores con 0: 0 Valores totales: 307641 Ceros del total: 0.0%
            
SUPPLIER Valores con 0: 0 Valores totales: 307641 Ceros del total: 0.0%
            
ITEM DESCRIPTION Valores con 0: 0 Valores totales: 307641 Ceros del total: 0.0%
            
ITEM TYPE Valores con 0: 0 Valores totales: 307641 Ceros del total: 0.0%
            
RETAIL SALES Valores con 0: 121701 Valores totales: 307641 Ceros del total: 39.55942153354072%
            
RETAIL TRANSFERS Valores con 0: 188460 Valores totales: 307641 Ceros del total: 61.259715057485835%
            
WAREHOUSE SALES Valores con 0: 96947 Valores totales: 307641 Ceros del total: 31.51302979771877%
            


In [336]:
for i in X_train.columns:
    x=np.sum( X_train[i] == 0.000)
    print(f'''{i} Valores con 0: {x} Valores totales: { X_train[i].count()} Ceros del total: {x*100/ X_train[i].count()}%
            ''')

YEAR Valores con 0: 0 Valores totales: 230731 Ceros del total: 0.0%
            
MONTH Valores con 0: 0 Valores totales: 230731 Ceros del total: 0.0%
            
SUPPLIER Valores con 0: 0 Valores totales: 230731 Ceros del total: 0.0%
            
ITEM DESCRIPTION Valores con 0: 0 Valores totales: 230731 Ceros del total: 0.0%
            
ITEM TYPE Valores con 0: 0 Valores totales: 230731 Ceros del total: 0.0%
            
RETAIL TRANSFERS Valores con 0: 141338 Valores totales: 230731 Ceros del total: 61.25661484585947%
            
WAREHOUSE SALES Valores con 0: 72642 Valores totales: 230731 Ceros del total: 31.483415752543003%
            


In [337]:
for i in X_test.columns:
    x=np.sum(X_test[i] == 0.000)
    print(f'''{i} Valores con 0: {x} Valores totales: {X_test[i].count()} Ceros del total: {x*100/X_test[i].count()}%
            ''')

YEAR Valores con 0: 0 Valores totales: 46146 Ceros del total: 0.0%
            
MONTH Valores con 0: 0 Valores totales: 46146 Ceros del total: 0.0%
            
SUPPLIER Valores con 0: 0 Valores totales: 46146 Ceros del total: 0.0%
            
ITEM DESCRIPTION Valores con 0: 0 Valores totales: 46146 Ceros del total: 0.0%
            
ITEM TYPE Valores con 0: 0 Valores totales: 46146 Ceros del total: 0.0%
            
RETAIL TRANSFERS Valores con 0: 28324 Valores totales: 46146 Ceros del total: 61.379101113856024%
            
WAREHOUSE SALES Valores con 0: 14548 Valores totales: 46146 Ceros del total: 31.526026091102153%
            


In [338]:
for i in X_val.columns:
    x=np.sum(X_val[i] == 0.000)
    print(f'''{i} Valores con 0: {x} Valores totales: {X_val[i].count()} Ceros del total: {x*100/X_val[i].count()}%
            ''')

YEAR Valores con 0: 0 Valores totales: 30764 Ceros del total: 0.0%
            
MONTH Valores con 0: 0 Valores totales: 30764 Ceros del total: 0.0%
            
SUPPLIER Valores con 0: 0 Valores totales: 30764 Ceros del total: 0.0%
            
ITEM DESCRIPTION Valores con 0: 0 Valores totales: 30764 Ceros del total: 0.0%
            
ITEM TYPE Valores con 0: 0 Valores totales: 30764 Ceros del total: 0.0%
            
RETAIL TRANSFERS Valores con 0: 18798 Valores totales: 30764 Ceros del total: 61.10388766090235%
            
WAREHOUSE SALES Valores con 0: 9757 Valores totales: 30764 Ceros del total: 31.715641659082046%
            


# BUSCAR CORRECLACIONES, ELIMINAR ELEMENTOS QUE NO IMPLEMENTAN VALOR.................